interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp

```diff
@@ -0,0 +1,302 @@
+                                                       fromASTContext,
+                                                       m_ChildInterpreter,
+                                                       m_ParentInterpreter,
+                                                       *this);
+
+    for (DeclContext::lookup_iterator I = lookup_result.begin(),
+          E = lookup_result.end(); I != E; ++I) {
+      // Check if this Name we are looking for is
+      // a DeclContext (for example a Namespace, function etc.).
+      if (DeclContext *declContextToImport = llvm::dyn_cast<DeclContext>(*I)) {
+
+        ImportDeclContext(declContextToImport, importer, childDeclName,
+                          parentDeclName, childCurrentDeclContext);
+
+      } else if (Decl *declToImport = llvm::dyn_cast<Decl>(*I)) {
+        // else it is a Decl
+        ImportDecl(declToImport, importer, childDeclName,
+                   parentDeclName, childCurrentDeclContext);
+      }
+    }
+    return true;
+  }
+
+  ///\brief This is the one of the most important function of the class
+  /// since from here initiates the lookup and import part of the missing
+  /// Decl(s) (Contexts).
+  ///
+  bool ExternalInterpreterSource::FindExternalVisibleDeclsByName(
+    const DeclContext *childCurrentDeclContext, DeclarationName childDeclName) {
+
+    assert(childDeclName && "Child Decl name is empty");
+
+    assert(childCurrentDeclContext->hasExternalVisibleStorage() &&
+           "DeclContext has no visible decls in storage");
+
+    //Check if we have already found this declaration Name before
+    DeclarationName parentDeclName;
+    std::map<clang::DeclarationName,
+             clang::DeclarationName>::iterator IDecl =
+                                            m_ImportedDecls.find(childDeclName);
+    if (IDecl != m_ImportedDecls.end()) {
+      parentDeclName = IDecl->second;
+    } else {
+      // Get the identifier info from the parent interpreter
+      // for this Name.
+      std::string name = childDeclName.getAsString();
+      IdentifierTable &parentIdentifierTable =
+                            m_ParentInterpreter->getCI()->getASTContext().Idents;
+      IdentifierInfo &parentIdentifierInfo =
+                            parentIdentifierTable.get(name);
+      parentDeclName = DeclarationName(&parentIdentifierInfo);
+    }
+
+    // Search in the map of the stored Decl Contexts for this
+    // Decl Context.
+    std::map<const clang::DeclContext *, clang::DeclContext *>::iterator
+          IDeclContext = m_ImportedDeclContexts.find(childCurrentDeclContext);
+    // If childCurrentDeclContext was found before and is already in the map,
+    // then do the lookup using the stored pointer.
+    if (IDeclContext == m_ImportedDeclContexts.end()) return false;
+
+    DeclContext *parentDeclContext = IDeclContext->second;
+
+    ASTContext &fromASTContext =
+          Decl::castFromDeclContext(parentDeclContext)->getASTContext();
```

Comment:
These are fairly expensive operations. Maybe we should set the to and from `ASTContext`s in the constructor of the `ExternalInterpreterSource`?